### PR TITLE
[FLINK-5714] [table] Use a builder pattern for creating CsvTableSource

### PR DESCRIPTION
**Pull request by wuchong**
Thanks for contributing to Apache Flink. Before you open your pull request, please take the following check list into consideration.
If your changes take all of the items into account, feel free to open your pull request. For more information and/or questions please refer to the [How To Contribute guide](http://flink.apache.org/how-to-contribute.html).
In addition to going through the list, please provide a meaningful description of your changes.

- [x] General
  - The pull request references the related JIRA issue ("[FLINK-XXX] Jira title text")
  - The pull request addresses only one issue
  - Each commit in the PR has a meaningful commit message (including the JIRA id)

- [x] Documentation
  - Documentation has been added for new functionality
  - Old documentation affected by the pull request has been updated
  - JavaDoc for public methods has been added

- [x] Tests & Build
  - Functionality added by the pull request is covered by tests
  - `mvn clean verify` has been executed successfully locally or a Travis build has passed

Add a builder to create CsvTableSource. And I have also updated the documentation.

### CODE REVIEW

File: docs/dev/table_api.md [hunk 1]
```diff
@@ -253,56 +253,51 @@ Table result = tableEnvironment.ingest("kafka-source");
...
- - `ignoreComments` An optional prefix to indicate comments, `null` by default.
- - `lenient` Flag to skip records with parse error instead to fail, `false` by default.
+It can be created by `CsvTableSource.builder()`, the builder has the following methods to configure properties:
+
+ - `path(String path)` Sets the path to the CSV file, required.
+ - `field(String fieldName, TypeInformation<?> fieldType)` Adds a field with the field name and field type information, required.
+ - `fieldDelimiter(String delim)` Sets the field delimiter, `","` by default.
+ - `lineDelimiter(String delim)` Sets the line delimiter, `"\n"` by default.
+ - `quoteCharacter(Character quote)` Sets the quote character for String values, `null` by default.
+ - `commentPrefix(String prefix)` Sets a prefix to indicate comments, null by default.
```
Comment (KurtYoung) [inline docs/dev/table_api.md]: Shall we change to null to `null` to keep consistency

File: docs/dev/table_api.md [hunk 2]
```diff
@@ -253,56 +253,51 @@ Table result = tableEnvironment.ingest("kafka-source");
...
+ - `ignoreFirstLine()` Ignore the first line. Not skip the first line by default.
```
Comment (KurtYoung) [inline docs/dev/table_api.md]: disabled by default?

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/sources/CsvTableSource.scala [hunk 3]
```diff
@@ -138,4 +141,160 @@ class CsvTableSource(
...
+
+object CsvTableSource {
+
+  /**
+    * A builder for creating [[CsvTableSource]] instances.
+    *
+    * For example:
+    *
+    * {{{
+    *   val source: CsvTableSource = new CsvTableSourceBuilder()
```
Comment (KurtYoung) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/sources/CsvTableSource.scala]: CsvTableSourceBuilder should be CsvTableSource.builder() 
and does this comment a little duplicated with method builder ?

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/sources/CsvTableSource.scala [hunk 4]
```diff
@@ -138,4 +141,160 @@ class CsvTableSource(
...
+    *     .path("/path/to/your/file.csv")
+    *     .field("myfield", Types.STRING)
+    *     .field("myfield2", Types.INT)
+    *     .build()
+    * }}}
+    *
+    */
+  class Builder {
+
+    private val fieldNames: ListBuffer[String] = ListBuffer[String]()
```
Comment (KurtYoung) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/sources/CsvTableSource.scala]: Can we bind the name and type to a MutableMap and do some check when we adding fields? Like can not have duplicated filed names?

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/sources/CsvTableSource.scala [hunk 5]
```diff
@@ -44,15 +47,15 @@ import org.apache.flink.table.api.TableException
...
-    path: String,
-    fieldNames: Array[String],
-    fieldTypes: Array[TypeInformation[_]],
-    fieldDelim: String = CsvInputFormat.DEFAULT_FIELD_DELIMITER,
-    rowDelim: String = CsvInputFormat.DEFAULT_LINE_DELIMITER,
-    quoteCharacter: Character = null,
-    ignoreFirstLine: Boolean = false,
-    ignoreComments: String = null,
-    lenient: Boolean = false)
+    private val path: String,
```
Comment (KurtYoung) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/sources/CsvTableSource.scala]: I think these two forms are identical

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/sources/CsvTableSource.scala [hunk 6]
```diff
@@ -138,4 +141,160 @@ class CsvTableSource(
...
+      this.lenient = true
+      this
+    }
+
+    /**
+      * Apply the current values and constructs a newly-created [[CsvTableSource]].
+      * @return a newly-created [[CsvTableSource]].
+      */
+    def build: CsvTableSource = {
+      Preconditions.checkNotNull(path, "Path must not be null.")
```
Comment (KurtYoung) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/sources/CsvTableSource.scala]: Fields can not be empty too?

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/sources/CsvTableSource.scala (hunk 3 above)
Comment (wuchong) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/sources/CsvTableSource.scala]: Yes, `new CsvTableSourceBuilder()` should be `new CsvTableSource.Builder()`. The comment is a little duplicate. But I think it is reasonable to let user create a builder by the constructor, not only the static `builder` method.

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/sources/CsvTableSource.scala (hunk 4 above)
Comment (wuchong) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/sources/CsvTableSource.scala]: Good point!

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/sources/CsvTableSource.scala (hunk 5 above)
Comment (wuchong) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/sources/CsvTableSource.scala]: Actually, I want to access these fields in the `equals(Object)` by `that.path`. The `private val` modifier will create getter/setter implicitly.

**Comment by wuchong** [comment]
Thank you for your advice. I will update the PR soon.

**Comment by wuchong** [comment]
updated!

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/sources/CsvTableSource.scala [hunk 7]
```diff
@@ -266,11 +266,16 @@ object CsvTableSource {
       * @return a newly-created [[CsvTableSource]].
       */
     def build: CsvTableSource = {
-      Preconditions.checkNotNull(path, "Path must not be null.")
+      if (path == null) {
+        throw new IllegalArgumentException("Path must be defined.")
```
Comment (KurtYoung) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/sources/CsvTableSource.scala]: You can still use Preconditions.checkArgument though.

**Comment by twalthr** [comment]
Looks good to merge. Merging...

**Comment by twalthr** [comment]
I found another issue: We should use a `LinkedHashMap` otherwise the order of fields and types is not guaranteed. I fixed it and will merge this now.

**Pull request closed without merging**